File: leetcode/0069-sqrtx/solution.py
class Solution:
    def mySqrt(self, x: int) -> int:
        lo, hi = 0, x
        while lo < hi:
            mid = (lo + hi + 1) // 2
            if mid * mid <= x:
                lo = mid
            else:
                hi = mid - 1
        return lo


if __name__ == "__main__":
    print(Solution().mySqrt(4))  # 2
    print(Solution().mySqrt(8))  # 2
    print(Solution().mySqrt(0))  # 0

File: leetcode/0069-sqrtx/solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int mySqrt(int x) {
        long long lo = 0, hi = x;
        while (lo < hi) {
            long long mid = (lo + hi + 1) / 2;
            if (mid * mid <= x) lo = mid;
            else hi = mid - 1;
        }
        return (int)lo;
    }
};

int main() {
    Solution sol;
    cout << sol.mySqrt(4) << endl;   // 2
    cout << sol.mySqrt(8) << endl;   // 2
    cout << sol.mySqrt(0) << endl;   // 0
    return 0;
}